src/python/pants/option/parser.py

```diff
@@ -788,30 +790,44 @@ def record_option(value, rank, option_details=None):
 
         # Helper function to check various validity constraints on final option values.
         def check(val):
-            if val is not None:
-                choices = kwargs.get("choices")
-                if choices is None and "type" in kwargs:
-                    type_arg = kwargs.get("type")
-                    if inspect.isclass(type_arg) and issubclass(type_arg, Enum):
-                        choices = list(type_arg)
-                # TODO: convert this into an enum() pattern match!
-                if choices is not None and val not in choices:
-                    raise ParseError(
-                        "`{}` is not an allowed value for option {} in {}. "
-                        "Must be one of: {}".format(val, dest, self._scope_str(), choices)
-                    )
-                elif kwargs.get("type") == dir_option and not os.path.isdir(val):
-                    raise ParseError(
-                        "Directory value `{}` for option {} in {} does not exist.".format(
-                            val, dest, self._scope_str()
-                        )
-                    )
-                elif kwargs.get("type") == file_option and not os.path.isfile(val):
-                    raise ParseError(
-                        "File value `{}` for option {} in {} does not exist.".format(
-                            val, dest, self._scope_str()
-                        )
-                    )
+            if val is None:
+                return
+            choices = kwargs.get("choices")
+            type_arg = kwargs.get("type")
+            if choices is None and "type" in kwargs:
+                if inspect.isclass(type_arg) and issubclass(type_arg, Enum):
+                    choices = list(type_arg)
+            # TODO: convert this into an enum() pattern match!
+            if choices is not None and val not in choices:
+                raise ParseError(
+                    "`{}` is not an allowed value for option {} in {}. "
+                    "Must be one of: {}".format(val, dest, self._scope_str(), choices)
+                )
+
+            if type_arg == file_option:
+                check_file_exists(val)
+            if type_arg == dir_option:
+                check_dir_option(val)
```

Comment:
I extracted these out in part so that it's easier to start validating the values of a list. Currently, we only validate the `type` and not the `member_type`. So, you can have a list of file options where those files don't actually exist.